### PR TITLE
add/strict

### PR DESCRIPTION
**Pull request by odersky**
Add -strict option to  do some type checks that are encessary to ensure type soundness, but
are stricter than what Scala 2.x enforces.

The first such test is the "pattern cannot be uniquely instantiated" problem where we reject a
non-variant case subclass of a covariant superclass in a pattern match. The error is now only
issued in -struct mode, otherwise it will be a warning.

We might move more tests into the same category. This should help the transition. Review by @DarkDimius

### CODE REVIEW

File: src/dotty/tools/dotc/config/ScalaSettings.scala [hunk 1]
```diff
@@ -39,6 +39,7 @@ class ScalaSettings extends Settings.SettingGroup {
   val defines = PrefixSetting("-Dproperty=value", "-D", "Pass -Dproperty=value directly to the runtime system.")
   val toolcp = PathSetting("-toolcp", "Add to the runner classpath.", "")
   val nobootcp = BooleanSetting("-nobootcp", "Do not use the boot classpath for the scala jars.")
+  val strict = BooleanSetting("-strict", "Use strict type rules, which means some formerly legal code does not typecheck anymore.")
```
Comment (xeno-by) [inline src/dotty/tools/dotc/config/ScalaSettings.scala]: Is there a document that describes what Scala code can be used with Dotty without any changes and what requires migration?

Comment (odersky) [inline src/dotty/tools/dotc/config/ScalaSettings.scala]: Not yet. I think we should work one having one though.

**Comment by smarter** [comment]
Shouldn't we run the tests with -strict?

**Comment by DarkDimius** [comment]
Codewise LGTM,
But, @odersky, what should be the default? 
Does this mean that we want to be unsoundness-compatible with scala?
I'd better say that there should flag that allows to be unsound, not the one that enables soundness.

**Comment by odersky** [comment]
For the moment I would prioritize scala compatibility, also for the tests. Once we have everything under dotty we can selectively enforce -strict.

**Pull request merged**